cachi2/core/package_managers/rpm/main.py

```diff
@@ -0,0 +1,204 @@
+            file_path,
+        ]
+        epoch = run_cmd(cmd=["rpm", *rpm_args], params={})
+
+        # NOTE: add license into SBOM?
+        rpm_args = [
+            "-q",
+            "--queryformat",
+            "%{LICENSE}",
+            file_path,
+        ]
+        license = run_cmd(cmd=["rpm", *rpm_args], params={})
+
+        with rpmfile.open(file_path) as rpm:
+            vendor = rpm.headers.get("vendor", b"").decode()
+            vendor = quote(vendor.lower())
+
+            name = rpm.headers.get("name", b"").decode()
+            version = rpm.headers.get("version", b"").decode()
+            release = rpm.headers.get("release", b"").decode()
+            arch = rpm.headers.get("arch", b"").decode()
+            download_url = quote(file_metadata["url"])
+
+            # https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm
+            # https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairsa
+            purl = (
+                f"pkg:rpm{'/' if vendor else ''}{vendor}/{name}@{version}{release}"
+                f"?arch={arch}{'&epoch=' if epoch else ''}{epoch}&download_url={download_url}"
+            )
+
+            components.append(
+                Component(
+                    name=name,
+                    version=version,
+                    purl=purl,
+                    # NOTE: add checksum into SBOM?
+                )
+            )
+
+
+def process_packages(path: Path) -> None:
+    """
+    Search structure for all repoid dirs and create repository metadata
+    out of its RPMs. Skip sources dir that contains only SRPMs.
+    """
+    package_dir = RootedPath(path).join_within_root(DEFAULT_PACKAGE_DIR)
+    for entry in os.listdir(package_dir):
+        if entry == "sources" or not os.path.isdir(os.path.join(package_dir, entry)):
+            continue
+        arch = entry
+        for entry in os.listdir(os.path.join(package_dir, arch)):
+            if not os.path.isdir(os.path.join(package_dir, arch, entry)):
+                continue
+            repoid = entry
+            repodir = os.path.join(package_dir, arch, repoid)
+            createrepo(repoid, repodir)
+
+
+def createrepo(repoid: str, repodir: str) -> None:
+    log.info(f"Creating repository metadata for repoid '{repoid}': {repodir}")
+    cmd = ["/usr/bin/createrepo", repodir]
+    log.debug("$ " + shlex.join(cmd))
+    stdout = run_cmd(cmd, params={})
+    log.debug(stdout)
+
+
+def _generate_repofiles(path: RootedPath) -> list[ProjectFile]:
+    """
+    Search structure and generate repofile for each arch in its dir.
+    Repofile contains all arch's repoids. Skip sources dir that contains only SRPMs.
+    """
+    project_files: list[ProjectFile] = []
+    package_dir = path.join_within_root(DEFAULT_PACKAGE_DIR)
+    for subdir in os.listdir(package_dir):
+        if subdir == "sources":
```

Comment:
Why are we calling this directory 'sources' when the common name among RPM distros is just 'source'?

Reply:
_sources_ vs. _source_. Actually, I don't know. _sources_ was in the design from the beginning. It would probably look better or would be more unified with the current public repository structures. But I am afraid to change it now. I think it would require changing not just the directory name, but also lockfile section name which would mean spreading this information and changing the existing testing data across the pipeline. The section name in lockfile and output directory are two things but having them different would look weird to me and possibly confusing to others. Does the _sources_ break anything?

Reply:
Yes, in the thread you link sources and packages are stored together within common architecture.
I discussed it further and the result was published here:
https://github.com/containerbuildsystem/cachi2/discussions/465#discussioncomment-8747459
including the reasons.

Reply:
Citing from #465:
> - Splitting sources per arches and repoids. Despite we will probably deliver single source image (SRPMs might eventually be in one common directory) I prefer not to lose the information about the structure and it is more general for other eventual use-cases.

Unrelated to the matter at hand.

> - separate all sources to their own structure so createrepo doesn't have to process them when together with rpms. And when traversing the sources structure it is not so large.

I honestly don't understand ^this comment at all. Source RPMs are still RPMs belonging to a some repository. I don't understand why we're treating them differently. Again, hitting the issue of design being tailored to a specific use case. The fact that some project/product doesn't need to create a repository out of SRPMs and only needs them in the form of plain files to stuff them to some source container is irrelevant to cachi2 as a tool. We should treat all types of RPMs consistently, i.e. if they belong to a repository denoted by a `repoid` we need to put them along other repositories. Note that official repositories will clearly mark such repoids, e.g. `fedora-source`. Also note that `createrepo` and generation of the `.repo` file isn't something that the use case you're hinting at (i.e. creating source containers) will do, so the argument about not having to process SRPMs together with RPMs IMO doesn't stand. Quite the contrary, once you get to the use case of actually wanting a `.repo` file created by cachi2 you want a repository that contains everything the lockfile specifies which currently isn't the case since you keep the sources separate and omit them from the repo file generation - that is an inconsistent and non-intuitive approach.

Reply:
I think changing the output structure is quite on us while changing the lockfile format (section sources --> source) has a wider impact on teams who could already generate lockfiles.
Technically, we can have the directory name and the lockfile section different, but I would preferably keep them both unchanged.

Reply:
> Technically, we can have the directory name and the lockfile section different, but I would preferably keep them both unchanged.

Regardless of the lockfile format and contents, why is directory name being consistent with the lockfile so important? Cachi2 decides what the output looks like and I think consistency is key here which the placement of the source directory as proposed doesn't comply with. Changing other parts of the consuming pipeline to match this should be pretty straightforward while keeping the lockfile format unchanged.

Reply:
Ok then. I will rename both.

Reply:
```
<output>/deps/rpm/<arch>/<repoid>/*.rpm
<output>/deps/rpm/<arch>/cachi2.repo OR <output>/deps/rpm/<arch>/repos.d/cachi2.repo
<output>/deps/rpm/<arch>/<repoid-source>/*.src.rpm
```
Let's have the structure above. OK. I will face the situation <repoid-source> will evaluate like "fedora-source-source". I think technically it is not an issue, it just looks weird.

Reply:
> I will face the situation will evaluate like "fedora-source-source".

Why? I didn't mean to suggest to use "fedora-source-source" with `<repoid-source>` I only needed to visually separate the placeholder from plain `<repoid>`. So what I really expect in real world would be the exact repoid that the lockfile specifies for source repos, i.e. `fedora-source` being the name of the actual directory in this case.

Reply:
Ok, I understood that you meant the "-source" suffix. Then I will have to generate repository metadata for every repoid including repoids for sources via `createrepo`. Because I don't know how to distinguish package repoid and source repoid. I think it is not a big issue - metadata for sources repoids just won't have any usage.

On the other hand, I will be able not to include source repoids into repofiles.

Reply:
OK, then. Let's have source repoids included in the repofile.